### PR TITLE
configure resource-aggregate service via yaml file

### PR DESCRIPTION
**Pull request by jkralik**

### CODE REVIEW

**Comment by ondrejtomcik** [comment]
> Co-authored-by: Trishia Yun [trishia1298@imsoftlab.com](mailto:trishia1298@imsoftlab.com)

This belongs to merge commit message. not here.

**Comment by jkralik** [comment]
> 
> 
> > Co-authored-by: Trishia Yun [trishia1298@imsoftlab.com](mailto:trishia1298@imsoftlab.com)
> 
> This belongs to merge commit message. not here.

Yes, it is there -> it was made from commit.

File: authorization/config.yaml [hunk 1]
```diff
@@ -2,14 +2,14 @@ log:
   debug: false
 apis:
   grpc:
-    address: "0.0.0.0:9000"
+    address: "0.0.0.0:9100"
```
Comment (ondrejtomcik) [inline authorization/config.yaml:5]: why are you changing ports in the config refactoring branch?

File: resource-aggregate/config.yaml [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  debug: false
+apis:
+  grpc:
+    address: "0.0.0.0:9100"
+    tls:
+      caPool: "/secrets/public/rootca.crt"
+      keyFile: "/secrets/private/cert.key"
+      certFile: "/secrets/public/cert.crt"
+      clientCertificateRequired: true
+    capabilities:
```
Comment (ondrejtomcik) [inline resource-aggregate/config.yaml]: what has snapshotTreshold and occ to do with grpc capabilities?

File: resource-aggregate/config.yaml [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+      snapshotThreshold: 16
+      occMaxRetry: 8
+      userMgmtTickFrequency: 15s
```
Comment (ondrejtomcik) [inline resource-aggregate/config.yaml]: What are these userMgmt configuration values about?

File: resource-aggregate/config.yaml [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+  eventStore:
+    mongoDB:
+      uri:
+      database: eventStore
+      tls:
+        caPool: "/secrets/public/rootca.crt"
+        keyFile: "/secrets/private/cert.key"
+        certFile: "/secrets/public/cert.crt"
+        useSystemCAPool: false
+      batchSize: 16
```
Comment (ondrejtomcik) [inline resource-aggregate/config.yaml]: description?

File: resource-aggregate/config.yaml [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+      maxPoolSize: 16
```
Comment (ondrejtomcik) [inline resource-aggregate/config.yaml]: description?

File: resource-aggregate/config.yaml [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+      maxConnIdleTime: 4m0s
+  oauthProvider:
```
Comment (ondrejtomcik) [inline resource-aggregate/config.yaml]: resource directory doesn't have oauth providers. It might have oauth client for a specific API.

File: resource-aggregate/config.yaml [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    jwks:
```
Comment (ondrejtomcik) [inline resource-aggregate/config.yaml]: jwks shouldn't be configurable. There should be only 
```
authorization:
      authority: ""
      audience: ""
```
per API. by the way, is this part of the authorization service for each of its API (http and grpc)?

File: resource-aggregate/config.yaml [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+      tickFrequency: 10s
+    ownerClaim: sub
+  authorizationServer:
+    address:
+    tls:
+      caPool: "/secrets/public/rootca.crt"
+      keyFile: "/secrets/private/cert.key"
+      certFile: "/secrets/public/cert.crt"
+      useSystemCAPool: false
+    keepAlive:
```
Comment (ondrejtomcik) [inline resource-aggregate/config.yaml]: What is this authorization server keepalive?

File: resource-aggregate/cqrs/eventbus/nats/configV2.go [hunk 9]
```diff
@@ -0,0 +1,21 @@
+package nats
+
+import (
+	"fmt"
+
+	nats "github.com/nats-io/nats.go"
+	"github.com/plgd-dev/cloud/pkg/security/certManager/client"
+)
+
+type ConfigV2 struct {
```
Comment (ondrejtomcik) [inline resource-aggregate/cqrs/eventbus/nats/configV2.go:10]: why is not v2 part of the namespace / import but part of the struct?

File: resource-aggregate/service/service_test.go [hunk 10]
```diff
@@ -3,93 +3,66 @@ package service_test
...
 	_, err = raClient.PublishResourceLinks(ctx, pubReq)
 	require.NoError(t, err)
 
-	unpubReq := testMakeUnpublishResourceRequest(deviceId, href, resp.UserId, resp.AccessToken)
+	unpubReq := testMakeUnpublishResourceRequest(deviceId, []string{href})
 	_, err = raClient.UnpublishResourceLinks(ctx, unpubReq)
 	require.NoError(t, err)
 }
 
+/*
```
Comment (ondrejtomcik) [inline resource-aggregate/service/service_test.go]: Why there is a commented code?

File: resource-aggregate/service/service_test.go [hunk 11]
```diff
@@ -3,93 +3,66 @@ package service_test
...
 	})
 	require.NoError(t, err)
 
-	pubReq := testMakePublishResourceRequest(deviceId, href, resp.UserId, resp.AccessToken)
+	pubReq := testMakePublishResourceRequest(deviceId, []string{href})
 	_, err = raClient.PublishResourceLinks(ctx, pubReq)
 	require.NoError(t, err)
 
-	unpubReq := testMakeUnpublishResourceRequest(deviceId, href, resp.UserId, resp.AccessToken)
+	unpubReq := testMakeUnpublishResourceRequest(deviceId, []string{href})
```
Comment (ondrejtomcik) [inline resource-aggregate/service/service_test.go:60]: Why did you change in config refactoring args of test func ?

File: authorization/config.yaml (hunk 1 above)
Comment (jkralik) [reply authorization/config.yaml:5]: Default port was 9100

File: resource-aggregate/config.yaml (hunk 2 above)
Comment (jkralik) [reply resource-aggregate/config.yaml]: I want to change it:
```
clients:
  aggregate:
     snapshotTreshold:
     occ:
     eventStore:
 ```
 what do you think ?

File: resource-aggregate/config.yaml (hunk 3 above)
Comment (jkralik) [reply resource-aggregate/config.yaml]: we can change it:
```
clients:
  userMgmt:
    authServer:
    tickFrequency:
```

File: resource-aggregate/config.yaml (hunk 5 above)
Comment (jkralik) [reply resource-aggregate/config.yaml]: number of connection

File: resource-aggregate/service/service_test.go (hunk 11 above)
Comment (jkralik) [reply resource-aggregate/service/service_test.go:60]: I merged test functions.

**Comment by ondrejtomcik** [comment]
```yaml
log:
  debug: false
apis:
  grpc:
    address: "0.0.0.0:9100"
    authorization:
      authority: ""
      audience: ""
      http:
        maxIdleConns: 16
        maxConnsPerHost: 32
        maxIdleConnsPerHost: 16
        idleConnTimeout: 30s
        timeout: 10s
        tls:
          caPool: "/secrets/public/rootca.crt"
          keyFile: "/secrets/private/cert.key"
          certFile: "/secrets/public/cert.crt"
          useSystemCAPool: false
    tls:
      caPool: "/secrets/public/rootca.crt"
      keyFile: "/secrets/private/cert.key"
      certFile: "/secrets/public/cert.crt"
      clientCertificateRequired: true   
clients:
  eventBus:
    nats:
      url:
      tls:
        caPool: "/secrets/public/rootca.crt"
        keyFile: "/secrets/private/cert.key"
        certFile: "/secrets/public/cert.crt"
        useSystemCAPool: false
  eventStore:
    snapshotThreshold: 16
    occMaxRetry: 8
    mongoDB:
      uri:
      database: eventStore
      tls:
        caPool: "/secrets/public/rootca.crt"
        keyFile: "/secrets/private/cert.key"
        certFile: "/secrets/public/cert.crt"
        useSystemCAPool: false
      batchSize: 16
      maxPoolSize: 16
      maxConnIdleTime: 4m0s
  authorizationServer:
    ownerClaim: sub
    pullFrequency: 15s
    cacheExpiration: 1m
    grpc:
      address:
      tls:
        caPool: "/secrets/public/rootca.crt"
        keyFile: "/secrets/private/cert.key"
        certFile: "/secrets/public/cert.crt"
        useSystemCAPool: false
      keepAlive:
        Time: 10s
        Timeout: 20s
        permitWithoutStream: true
    oauth:
      clientID:
      clientSecret:
      scopes: []
      tokenURL:
      audience:
      tokenVerifyFrequency: 10s
      http:
        maxIdleConns: 16
        maxConnsPerHost: 32
        maxIdleConnsPerHost: 16
        idleConnTimeout: 30s
        timeout: 10s
        tls:
          caPool: "/secrets/public/rootca.crt"
          keyFile: "/secrets/private/cert.key"
          certFile: "/secrets/public/cert.crt"
          useSystemCAPool: false
```

**Comment by ondrejtomcik** [comment]
authorization service update
```yaml
log:
  debug: false
apis:
  grpc:
    address: "0.0.0.0:9000"
    tls:
      caPool: "/secrets/public/rootca.crt"
      keyFile: "/secrets/private/cert.key"
      certFile: "/secrets/public/cert.crt"
      clientCertificateRequired: true
    authorization:
      authority: ""
      audience: ""
      http:
        maxIdleConns: 16
        maxConnsPerHost: 32
        maxIdleConnsPerHost: 16
        idleConnTimeout: 30s
        timeout: 10s
        tls:
          caPool: "/secrets/public/rootca.crt"
          keyFile: "/secrets/private/cert.key"
          certFile: "/secrets/public/cert.crt"
          useSystemCAPool: false
  http:
    address: "0.0.0.0:9001"
    tls:
      caPool: "/secrets/public/rootca.crt"
      keyFile: "/secrets/private/cert.key"
      certFile: "/secrets/public/cert.crt"
      clientCertificateRequired: true
oauthClients:
  device:
    provider: "plgd"
    clientID: ""
    clientSecret: ""
    scopes: []
    authorizationURL: ""
    tokenURL: ""
    redirectURL: "" 
    http:
      maxIdleConns: 16
      maxConnsPerHost: 32
      maxIdleConnsPerHost: 16
      idleConnTimeout: 30s
      timeout: 10s
      tls:
        caPool: "/secrets/public/rootca.crt"
        keyFile: "/secrets/private/cert.key"
        certFile: "/secrets/public/cert.crt"
        useSystemCAPool: false
    ownerClaim: "sub"
  client:
    clientID: ""
    scopes: []
    authorizationURL: ""
    audience: ""
    redirectURL: ""
    http:
      maxIdleConns: 16
      maxConnsPerHost: 32
      maxIdleConnsPerHost: 16
      idleConnTimeout: 30s
      timeout: 10s
      tls:
        caPool: "/secrets/public/rootca.crt"
        keyFile: "/secrets/private/cert.key"
        certFile: "/secrets/public/cert.crt"
        useSystemCAPool: false
clients:
  storage:
    mongoDB:
      uri: "mongodb://localhost:27017"
      database: "usersDevices"
      tls:
        caPool: "/secrets/public/rootca.crt"
        keyFile: "/secrets/private/cert.key"
        certFile:  "/secrets/public/cert.crt"
        useSystemCAPool: false
```

File: resource-aggregate/service/service_test.go (hunk 10 above)
Comment (jkralik) [reply resource-aggregate/service/service_test.go]: removed

File: resource-aggregate/cqrs/eventbus/nats/configV2.go (hunk 9 above)
Comment (jkralik) [reply resource-aggregate/cqrs/eventbus/nats/configV2.go:10]: This is tmp state for now. After we migrate al services we will replace Config with Config2.

File: resource-aggregate/config.yaml (hunk 8 above)
Comment (jkralik) [reply resource-aggregate/config.yaml]: part of grpc client

File: resource-aggregate/config.yaml (hunk 7 above)
Comment (jkralik) [reply resource-aggregate/config.yaml]: moved to authorization.

File: bundle/Dockerfile [hunk 12]
```diff
@@ -116,5 +117,6 @@ ENV SDK_OAUTH_SCOPES=openid
 ENV OWNER_CLAIM="sub"
 ENV SERVICE_CLIENT_CONFIGURATION_CLOUDAUTHORIZATIONPROVIDER="plgd"
 ENV SERVICE_OAUTH_CLIENT_ID=test
+ENV SERVICE_OAUTH_CLIENT_SECRET=secret
```
Comment (ondrejtomcik) [inline bundle/Dockerfile:120]: Is this required? Before empty value was accepted.

File: pkg/security/jwt/validator/config.go [hunk 13]
```diff
@@ -0,0 +1,28 @@
+package validator
+
+import (
+	"fmt"
+
+	"github.com/plgd-dev/cloud/pkg/net/http/client"
+)
+
+type Config struct {
```
Comment (ondrejtomcik) [inline pkg/security/jwt/validator/config.go:9]: I don't understand this code structure. You have validator/config.go which contains configuration structure required to authorize APIs with validator and then you the same level you have validator.go which has validator struct containing this validator and openId configuration which is part of the validator.go. Unify your code style approach please.

File: pkg/security/oauth/manager/configv2.go [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+)
+
+type ConfigV2 struct {
+	ClientID      string        `yaml:"clientID" json:"clientID"`
+	ClientSecret  string        `yaml:"clientSecret" json:"clientSecret"`
+	Scopes        []string      `yaml:"scopes" json:"scopes"`
+	TokenURL      string        `yaml:"tokenURL" json:"tokenURL"`
+	Audience      string        `yaml:"audience" json:"audience"`
+	HTTP          client.Config `yaml:"http" json:"http"`
+	TickFrequency time.Duration `yaml:"tickFrequency" json:"tickFrequency"`
```
Comment (ondrejtomcik) [inline pkg/security/oauth/manager/configv2.go]: Tick won't tell me what is ticking. Describe the purpose of this and not how often is it ticking.

File: resource-aggregate/config.yaml [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+    address: "0.0.0.0:9100"
+    tls:
+      caPool: "/secrets/public/rootca.crt"
+      keyFile: "/secrets/private/cert.key"
+      certFile: "/secrets/public/cert.crt"
+      clientCertificateRequired: true
+    authorization:
+      authority:
+      audience:
+      ownerClaim: sub
```
Comment (ondrejtomcik) [inline resource-aggregate/config.yaml:14]: this value is authorization service client related, it has nothing to do with authorization of the grpc endpoint of the resource aggregate. It's relevant for northbound interface, to AuthService.

File: resource-aggregate/config.yaml [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+  eventStore:
+    mongoDB:
+      uri:
+      database: eventStore
+      tls:
+        caPool: "/secrets/public/rootca.crt"
+        keyFile: "/secrets/private/cert.key"
+        certFile: "/secrets/public/cert.crt"
+        useSystemCAPool: false
+      # limits number of queries in one find request.
```
Comment (ondrejtomcik) [inline resource-aggregate/config.yaml:43]: what is one find request?

File: resource-aggregate/config.yaml [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+  authorizationServer:
+    grpc:
+      address:
+      tls:
+        caPool: "/secrets/public/rootca.crt"
+        keyFile: "/secrets/private/cert.key"
+        certFile: "/secrets/public/cert.crt"
+        useSystemCAPool: false
+      keepAlive:
+        Time: 10s
```
Comment (ondrejtomcik) [inline resource-aggregate/config.yaml]: camelCase

File: resource-aggregate/config.yaml [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+        Timeout: 20s
+        permitWithoutStream: true
+    pullFrequency: 15s
```
Comment (ondrejtomcik) [inline resource-aggregate/config.yaml]: please can you affect order/indentation in print? Is it based on the order in struct? Then move it to the top as it's not readable - clear to which parent it belongs to.

File: resource-aggregate/config.yaml [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+      # limits number of queries in one find request.
+      batchSize: 128
+      # limits number of connections.
+      maxPoolSize: 16
+      # close connection when idle time reach the value.
+      maxConnIdleTime: 4m0s
+    # tries to create the snapshot event after n events
+    snapshotThreshold: 16
+    # limits number of try to store event
+    occMaxRetry: 8
```
Comment (ondrejtomcik) [inline resource-aggregate/config.yaml]: please can you affect order/indentation in print? Is it based on the order in struct? Then move it to the top as it's not readable - clear to which parent it belongs to.

File: resource-aggregate/config.yaml [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+        maxConnsPerHost: 32
+        maxIdleConnsPerHost: 16
+        idleConnTimeout: 30s
+        timeout: 10s
+        tls:
+          caPool: "/secrets/public/rootca.crt"
+          keyFile: "/secrets/private/cert.key"
+          certFile: "/secrets/public/cert.crt"
+          useSystemCAPool: false
+      tickFrequency: 10s
```
Comment (ondrejtomcik) [inline resource-aggregate/config.yaml]: please can you affect order/indentation in print? Is it based on the order in struct? Then move it to the top as it's not readable - clear to which parent it belongs to.

Comment (ondrejtomcik) [inline resource-aggregate/config.yaml]: again, describe purpose. Nobody knows what this configuration affects. Users just knows that it affects the execution of some function...

`verifyServiceTokenFrequency: 10s`

File: resource-aggregate/cqrs/eventbus/nats/configV2.go (hunk 9 above)
Comment (ondrejtomcik) [reply resource-aggregate/cqrs/eventbus/nats/configV2.go:10]: Why do we need Config2 and cannot stay on Config? There is no backwards compatibility with plgd cloud1. You can rename it right away to Config.

File: authorization/config.yaml [hunk 21]
```diff
@@ -2,47 +2,16 @@ log:
...
-    client:
-      clientID: ""
-      scopes: []
-      authorizationURL: ""
-      audience: ""
-      redirectURL: ""
+    authorization:
+      authority:
+      audience:
+      ownerClaim: sub
```
Comment (ondrejtomcik) [inline authorization/config.yaml:14]: There will be also an HTTP endpoint to delete the device, or also list all user's devices. What's your argument then to have ownerClaim part of the grpc API?

It's the configuration of the service, how it organizes the data. It has nothing to do with the token validation which API does. It belongs to the core configuration of the service, not to API configuration as there might be 1..n of them.

It belongs to oauthClients.device

File: resource-aggregate/cqrs/eventbus/nats/configV2.go (hunk 9 above)
Comment (jkralik) [reply resource-aggregate/cqrs/eventbus/nats/configV2.go:10]: It will be renamed in RD merge request.

File: bundle/Dockerfile (hunk 12 above)
Comment (jkralik) [reply bundle/Dockerfile:120]: Yes now it is required.

File: authorization/config.yaml (hunk 21 above)
Comment (jkralik) [reply authorization/config.yaml:14]: Authorization works with JWT token and fill context with `OwnerID` so in code you just acquire OwnerID from context and not from JWT token. It's more clean.

File: resource-aggregate/config.yaml (hunk 16 above)
Comment (jkralik) [reply resource-aggregate/config.yaml:43]: When you get events (find function in mongo) from eventstore it can goes through parallel request.  And this it a limit how many query parameters will be in one get events(find for mongo).

File: pkg/security/jwt/validator/config.go (hunk 13 above)
Comment (jkralik) [reply pkg/security/jwt/validator/config.go:9]: WTF?

File: authorization/config.yaml (hunk 21 above)
Comment (ondrejtomcik) [reply authorization/config.yaml:14]: So do you want to introduce the same property also for HTTP API of this service? Because when this will be authorized, there will be endpoints to list all users devices, delete user device, and others. Based on your logic described above, this property will need to be there as ownerID is acquired from the context - HTTP context in this case.

File: pkg/security/jwt/validator/config.go (hunk 13 above)
Comment (ondrejtomcik) [reply pkg/security/jwt/validator/config.go:9]: Why is APIAuthorization config part of jwt/validator/config.go, OpenIdConfiguration part of validator.go together with another Validator struct? I don't understand the logic of your code organization.

File: authorization/config.yaml (hunk 21 above)
Comment (jkralik) [reply authorization/config.yaml:14]: HTTP API will be removed from authorization service.

File: resource-aggregate/config.yaml (hunk 15 above)
Comment (jkralik) [reply resource-aggregate/config.yaml:14]: Authorization works with JWT token and fill context with OwnerID so in code you just acquire OwnerID from context and not from JWT token. It's more clean.

File: authorization/config.yaml (hunk 21 above)
Comment (ondrejtomcik) [reply authorization/config.yaml:14]: It won't because user will be at some point able to share his devices with different users or even more important, delete his own device what is a HTTP request from the UI to the authorization service.

File: resource-aggregate/config.yaml (hunk 20 above)
Comment (jkralik) [reply resource-aggregate/config.yaml]: fixed

Comment (jkralik) [reply resource-aggregate/config.yaml]: fixed

File: pkg/security/oauth/manager/configv2.go (hunk 14 above)
Comment (jkralik) [reply pkg/security/oauth/manager/configv2.go]: fixed

File: pkg/security/jwt/validator/config.go (hunk 13 above)
Comment (jkralik) [reply pkg/security/jwt/validator/config.go:9]: OpenIdConfiguration was moved to separate package - `pkg/security/openid`

File: resource-aggregate/config.yaml (hunk 17 above)
Comment (jkralik) [reply resource-aggregate/config.yaml]: Fixed

File: resource-aggregate/config.yaml (hunk 18 above)
Comment (jkralik) [reply resource-aggregate/config.yaml]: Fixed

File: resource-aggregate/config.yaml (hunk 19 above)
Comment (jkralik) [reply resource-aggregate/config.yaml]: Fixed

File: authorization/config.yaml (hunk 21 above)
Comment (jkralik) [reply authorization/config.yaml:14]: So do you want to introduce the same property also for HTTP API of this service?  - yes

Comment (jkralik) [reply authorization/config.yaml:14]: It is part of same authorization as is in GRPC.

Comment (ondrejtomcik) [reply authorization/config.yaml:14]: That means it doesnt belong to the specific API. If you have it both you might have a conflict. This value is used as a identifier when assigning devices to owners in database. That’s why it doesn’t belong to API but is a general service cfg.

**Review by ondrejtomcik — APPROVED** [review]
